### PR TITLE
Minimal refresh of Solo5 public APIs

### PR DESCRIPTION
**Pull request by mato**
This is a minimal refresh of the Solo5 public APIs and their underlying backend implementations for ukvm, virtio and muen to fix the most egregious problems as described in #200.

Summary of high-level changes, see [solo5.h](kernel/solo5.h) for full details:
1. `solo5_app_main()`: Add `const` to `cmdline`, document as returning a status code.
2. `solo5_net_write_sync()`: Add `const` to `buf`, pass `size` as `size_t`, document and enforce failure cases.
3. `solo5_net_read_sync()`: Pass sizes as `size_t`, return `read_size` separately from error code, document and enforce failure cases.
4. `solo5_net_mac_str()`: Add `const` to return type.
5. `solo5_blk_write_sync()`: Add `const` to `buf`, use `size_t` for sizes, restrict to single-sector full writes, document and enforce failure cases.
6. `solo5_blk_read_sync()`: Use `size_t` for sizes, restrict to single-sector full reads, document and enforce failure cases.
7. `solo5_blk_sectors()`: Rename to `solo5_blk_capacity()` to lessen confusion with `solo5_blk_sector_size()`.
8. `solo5_console_write()`: Change return type to `void` and document as _not_ having explicit failure cases.
9. All other public functions: Clarify documentation.

No functional change intended, hopefully a lot of edge case bugs elided. Tests have been updated and are passing on Linux and FreeBSD for `ukvm` and `virtio`.

### CODE REVIEW

**Comment by mato** [comment]
@Kensan I've not tested this on Muen, you might want to double-check my changes.

**Comment by Kensan** [comment]
@mato Will give it a spin as soon as I can.

**Comment by Kensan** [comment]
@mato I tested the changes on Muen and the muen-net implementation requires some adaptation, see b424a29. With that **test_ping_serve** works again.

File: kernel/solo5.h [hunk 1]
```diff
@@ -31,69 +31,96 @@
...
+ * Writes (size) bytes from (*buf) at the sector (offset) to the block device.
+ * (size) must be equal to the sector size.
+ *
+ * On failure:
+ *     Returns -1. This includes invalid argument errors and incomplete writes.
+ * On success:
+ *     Returns 0.
  */
-int solo5_blk_write_sync(uint64_t sec, uint8_t *data, int n);
+int solo5_blk_write_sync(uint64_t offset, const uint8_t *buf, size_t size);
```
Comment (ricarkol) [inline kernel/solo5.h]: I can't remember the number of times I had to check if offset is in bytes or sectors. Could offset be something like `sector_t offset` or `sector_t sector`.

Also, size should be number of sectors. we might never support writing less than a sector. As that would require reading it first.

File: kernel/solo5.h [hunk 2]
```diff
@@ -136,8 +165,13 @@ uint64_t solo5_clock_monotonic(void);
...
+ * Blocks until monotonic time reaches until_nsecs or network I/O is possible,
+ * whichever is sooner.
+ *
+ * If network I/O is possible:
+ *     Returns 1.
+ * If monotonic time has reached until_nsecs and network I/O is not possible:
+ *     Returns 0.
  *
  * TODO: Extend this interface to select which I/O events are of interest.
  */
```
Comment (ricarkol) [inline kernel/solo5.h]: To be more like the POSIX poll(), the time should be a delta and not a future timestamp. Plus, that would remove this time reading in the mirage event loop:

```
        let timeout =
          match Time.select_next () with
          |None -> Time.Monotonic.(time () + of_nanoseconds 86_400_000_000_000L) (* one day = 24 * 60 * 60 s *)
          |Some tm -> tm
        in
        if poll timeout then begin
```

https://github.com/mirage/mirage-solo5/blob/master/lib/main.ml#L68

On the other hand, that's a risky change, so up to you.

File: ukvm/ukvm_module_blk.c [hunk 3]
```diff
@@ -79,7 +96,17 @@ static void hypercall_blkread(struct ukvm_hv *hv, ukvm_gpa_t gpa)
     ssize_t ret;
     off_t pos, end;
 
+#if 0
     assert(rd->len <= SSIZE_MAX);
+#else
+    /*
+     * XXX: Artificially limit to single sector writes for now.
```
Comment (hannesm) [inline ukvm/ukvm_module_blk.c]: `writes` -> `reads`

File: tests/test_blk/test_blk.c [hunk 4]
```diff
@@ -59,18 +55,17 @@ int check_sector_write(uint64_t sector)
...
 
     puts("\n**** Solo5 standalone test_blk ****\n\n");
 
     /*
      * Write and read/check one tenth of the disk.
      */
-    nsectors = solo5_blk_sectors();
-    for (i = 0; i <= nsectors; i += 10) {
+    nsectors = solo5_blk_capacity();
+    for (i = 0; i < nsectors; i += 10) {
```
Comment (hannesm) [inline tests/test_blk/test_blk.c]: here, `<=` is changed to `<`, is this intentional?

File: kernel/solo5.h [hunk 5]
```diff
@@ -102,15 +129,17 @@ int solo5_blk_rw(void);
...
 
 /*
- * Writes (n) bytes from (*buf) to the console. Returns number of bytes
- * actually written.
+ * Writes (size) characters from (*buf) to the console.
+ *
+ * Always succeeds, however the implementation may truncate messages (e.g. due
```
Comment (hannesm) [inline kernel/solo5.h]: this is not entirely true: if `write` returns `< 0`, an assertion is raised (in the ukvm implementation): https://github.com/mato/solo5/blob/13e7a06cfcb24452d0e8113057cf4719f7ca2915/ukvm/ukvm_core.c#L103

File: kernel/solo5.h [hunk 6]
```diff
@@ -136,8 +165,13 @@ uint64_t solo5_clock_monotonic(void);
 uint64_t solo5_clock_wall(void);
```
Comment (hannesm) [inline kernel/solo5.h]: would be great to specify error behaviour for these as well... at least the ukvm implementation asserts that `clock_gettime` returns non-zero

File: kernel/solo5.h (hunk 1 above)
Comment (mato) [inline kernel/solo5.h]: I believe that the use of "sectors" in this API comes from @djwillia's original MirageOS porting work; MirageOS uses sectors in it's block device [signatures](https://github.com/mirage/mirage-block/blob/master/src/mirage_block.mli#L68).

The current situation using a mix of sectors (for `offset`) and bytes (for `size`) is indeed confusing, however I'm not sure which of the two options is better in the long run:

1. Use byte offsets and sizes, and reject offsets and sizes that are not a multiple of the sector size.
2. Use sector offsets and sizes.

IMO in both cases there's no need for an explicit `sector_t`, `uint64_t` is fine.

In the long run I'd like to abstract block and network I/O over a common API, in which case we'd have to use bytes for offsets and sizes anyway. So, I'd prefer the first option.

Random sprinkling of Mirage folks for other opinions: @djs55 @gjaldon @samoht What do you think?

Comment (djs55) [inline kernel/solo5.h]: The MirageOS OCaml code uses sectors to avoid having to add a runtime error -- we're keen on eliminating runtime errors where we can. If solo5 becomes more like Unix and uses byte offsets then it's fine -- the OCaml code can perform the conversion as it already does for Unix.

File: kernel/solo5.h (hunk 2 above)
Comment (mato) [inline kernel/solo5.h]: In my experience, the use of relative timeouts in POSIX poll() leads to unnecessary complications in the common case for an event loop, which is when I/O events are regularly arriving _and_ the application has some timers scheduled. Ignoring the Mirage implementation for now, consider the two options in this scenario:

_If relative timeouts are used_: The application _must_ recalculate the relative _timeout_ passed to poll() in each iteration of the loop. Given that poll() does not return[1] the remaining time, this involves a call to get the current time.

_If absolute timeouts are used_: The application can keep calling poll(_t_) in each iteration of the loop, where _t_ is the nearest scheduled timer, until poll() indicates that the time _t_ has been reached[2], at which point the application selects the next scheduled timer.

[1] If poll() is implemented in the backend using select() then some implementations _do_ return the time not slept. However, POSIX permits either behaviour so in practice this cannot be relied upon.

[2] Our API does not currently clearly indicate if the timeout has expired. This is a bug. Here is a strawman proposal for a fixed interface (without any ability to select what I/O is interesting since we don't have any infrastructure for that yet):

````
/*
 * Wait until network packets are available or monotonic time reaches (deadline), whichever is sooner.
 *
 * On success:
 *    Returns 0, and (*io) is true if network packets are available, (*timed_out) is true if (deadline) has passed.
 * On failure:
 *    Returns -1.
 */
int solo5_wait(uint64_t deadline, bool *io, bool *timed_out);
````

In the above proposal I've deliberately renamed the call to `solo5_wait()` to disassociate it from the POSIX poll(); I've had several conversations with people who were confused by the "poll" in the name.

Note that the change in semantics (reporting whether or not timeout has expired) requires a bit more work from the (ukvm) backends. That's fine and can be optimized later using epoll()/timerfd_create() on Linux and kqueue() on FreeBSD.

In summary, I'd prefer that we stick with absolute timeouts for poll() and implement the above proposal or similar as part of this change.

Aside: I went hunting for some modern "wait for events" APIs while thinking about this, those [implemented](https://github.com/fuchsia-mirror/magenta/blob/master/docs/syscalls/waitset_wait.md) by Magenta (the kernel used in Google Fuchsia) are a useful source of future inspiration and incidentally exclusively use absolute timeouts.

File: tests/test_blk/test_blk.c (hunk 4 above)
Comment (mato) [inline tests/test_blk/test_blk.c]: Yes, the sector offset is 0-based, so valid values are (0 ... solo5_blk_capacity() - 1). Interestingly there are "bugs" in the current backends (e.g. virtio/bhyve and ukvm) where writing past the end of the "device" would succeed if the underlying device was backed by a file which is the case in all of our tests.

File: kernel/solo5.h (hunk 5 above)
Comment (mato) [inline kernel/solo5.h]: Thanks for pointing that out. The backend code could be improved to ignore _some_ errors that are obviously a sign of run-time resource exhaustion (`ENOSPC` comes to mind). For anything else it should probably still assert. In either case I don't see any point in returning a failure status to the guest; what could it reasonably do to "recover"?

File: kernel/solo5.h (hunk 6 above)
Comment (mato) [inline kernel/solo5.h]: `clock_gettime()` can fail due to (on Linux):
- `EFAULT` tp points outside the accessible address space.
- `EINVAL` The clk_id specified is not supported on this system.

We should never get `EFAULT` (since we've already verified the buffer passed by the guest). If we do then we want to terminate the guest with extreme prejudice anyway. We will never get `EINVAL` since `CLOCK_MONOTONIC` is supported on any sane system.

Comment (mato) [inline kernel/solo5.h]: ... in other words `solo5_clock_xxx()` can never fail. If they do for some reason then that is legitimately an unrecoverable (and fatal) error.

File: kernel/solo5.h (hunk 1 above)
Comment (mato) [inline kernel/solo5.h]: @djs55 Doesn't passing in a buffer that is not a multiple of the sector size count as a runtime error? The documentation in mirage_block.mli is silent about this other than insisting on the buffers being a multiple of the sector size. Likewise, what if an invalid sector is passed (beyond the end of the device)?

Comment (djs55) [inline kernel/solo5.h]: @mato yes that's right-- both of those cases will fail at runtime (possibly killing the app, not sure about that). We ought to define that behaviour a bit better! So perhaps it's more accurate to say that a runtime error exists (unfortunately) but there is one fewer case that can trigger it :) Maybe (big maybe) it would be possible to design the rest of errors out with clever use of types without being too annoying.

File: kernel/solo5.h (hunk 2 above)
Comment (ricarkol) [inline kernel/solo5.h]: Agree, going for `solo5_wait()` instead of `solo5_poll()` might be the easiest way to avoid confusion.

File: kernel/solo5.h (hunk 5 above)
Comment (hannesm) [inline kernel/solo5.h]: I'm with you that there's no way to recover from most `write` errors.  I'd rephrase `Always succeeds` to reflect the actual behaviour  (although the virtio backend using `outb` can never fail AFAICS).  Do you expect that compilations always have assertions enabled (and leading to fatal errors)?  Or should there be a call to `exit()` instead of `assert` (sorry if I missed some note which explains how to use assertions in Solo5)?

File: kernel/solo5.h (hunk 6 above)
Comment (hannesm) [inline kernel/solo5.h]: That is good to hear - related to above, maybe it makes sense to explicitly document in `solo5.h` that fatal errors may occur in each call, which lead to an assertion failure (or `exit`)!?

**Comment by mato** [comment]
Superseded by #245 which should address (or is in the process of addressing) the issues raised here.
